cylc/flow/task_pool.py

```diff
@@ -1477,304 +1448,524 @@ def can_spawn(self, name: str, point: 'PointBase') -> bool:
+            not transient
+            and itask.tdef.has_abs_triggers
+            and itask.state.prerequisites_are_not_all_satisfied()
         ):
-            itask.state.satisfy_me(self.abs_outputs_done)
+            itask.satisfy_me(
+                [f"{a[0]}/{a[1]}:{a[2]}" for a in self.abs_outputs_done]
+            )
+
+        if prev_flow_wait:
+            self._spawn_after_flow_wait(itask)
+            return None
+
+        self.db_add_new_flow_rows(itask)
+        return itask
+
+    def _spawn_after_flow_wait(self, itask: TaskProxy) -> None:
+        LOG.info(
+            f"spawning children of {itask.identity} after flow wait"
+        )
+        self.spawn_on_all_outputs(itask, completed_only=True)
+        # update flow wait status in the DB
+        itask.flow_wait = False
+        # itask.flow_nums = orig_fnums
+        self.workflow_db_mgr.put_update_task_flow_wait(itask)
+        return None
+
+    def _get_task_proxy(
+        self,
+        point: 'PointBase',
+        taskdef: 'TaskDef',
+        flow_nums: 'FlowNums',
+        flow_wait: bool = False,
+        transient: bool = False,
+        is_manual_submit: bool = False,
+        submit_num: int = 0
+    ) -> Optional['TaskProxy']:
+        """Spawn a task proxy and update its outputs from the DB. """
+
+        if not self.can_be_spawned(taskdef.name, point):
+            return None
+
+        itask = TaskProxy(
+            self.tokens,
+            taskdef,
+            point,
+            flow_nums,
+            flow_wait=flow_wait,
+            submit_num=submit_num,
+            transient=transient,
+            is_manual_submit=is_manual_submit
+        )
 
-        if flow_wait_done:
-            for outputs_str, fnums in (
-                self.workflow_db_mgr.pri_dao.select_task_outputs(
-                    itask.tdef.name, str(itask.point))
-            ).items():
+        if itask is not None:
+            # Update it with outputs that were already completed.
+            info = self.workflow_db_mgr.pri_dao.select_task_outputs(
+                itask.tdef.name, str(itask.point))
+            if not info:
+                self.db_add_new_flow_rows(itask)
+            spawn_kids = False
+            for outputs_str, fnums in info.items():
                 if flow_nums.intersection(fnums):
+                    if itask.flow_wait:
+                        spawn_kids = True
                     for msg in json.loads(outputs_str):
                         itask.state.outputs.set_completed_by_msg(msg)
-                    break
-            LOG.info(f"{itask} spawning on outputs after flow wait")
-            self.spawn_on_all_outputs(itask, completed_only=True)
-            return None
 
-        LOG.debug(f"[{itask}] spawned")
-        self.db_add_new_flow_rows(itask)
+            if spawn_kids:
+                self._spawn_after_flow_wait(itask)
+
         return itask
 
-    def force_spawn_children(
+    def set(  # noqa: A003
```

Comment:
I'm really not sure we want to noqa this. IMO as an internal interface, do we _need_ to use `set`?

Reply:
Ooh good spot.

Reply:
Could call it `set_outputs_or_prereqs`?

Reply:
Haha, I believe I did this at @oliver-sanders request!

Reply:
I guess you're no longer of that opinion @oliver-sanders ?

Reply:
Aah, it's a class method not a function so it can't conflict:

```python
def set():
    print('whoops, this is overriding the Python builtin "set" method')

>>> a = set()
whoops, this is overriding the Python builtin "set" method
```

```python
class Foo:
    def set():
        print('this is fine')

>>> a = set()
{}
```

Reply:
There is one way it can conflict, but it's pretty niche:

```python
>>> class Foo:
...     def set():
...         print('here')
...     ABC = set()
... 
here
```

Reply:
Presumably it would be better all around to just change the internal method name back to "set_blah" or whatever, and avoid the whole problem. I guess I took from your earlier review comments that it's nice to have the method name match the command name ... but not to the extent of shadowing the Python std lib.

Reply:
I think my comment was mostly about renaming away from `reset`, I'm happy with`set_prereqs_and_outputs`.

That said `set` is more obvious and clearer so I'm happy with this too. It can't cause any issues in this context.

Reply:
Changed it.